### ISSUE TITLE
occasional fail on unit.ClusteringBranch.Backfill

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#5182]
I am getting an occasional (about 1 in 10) failure on `unit.ClusteringBranch`, but only on my local machine.

```
Running 1 tests (output_dir: /Users/larkost/projects/rethinkdb-test/test/results/2015-12-08T13:50:35.9MPOGc)
Note: Google Test filter = ClusteringBranch.*
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from ClusteringBranch
[ RUN      ] ClusteringBranch.ReadWrite
[       OK ] ClusteringBranch.ReadWrite (175 ms)
[ RUN      ] ClusteringBranch.Backfill
src/unittest/clustering_branch.cc:184: Failure
Value of: mock_lookup(&store2, it->first)
  Actual: "6"
Expected: it->second
Which is: ""
[  FAILED  ] ClusteringBranch.Backfill (1295 ms)
[----------] 2 tests from ClusteringBranch (1470 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (1471 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ClusteringBranch.Backfill

 1 FAILED TEST
Traceback (most recent call last):
  File "/Users/larkost/projects/rethinkdb-test/test/common/test_framework.py", line 594, in run
    self.test.run()
  File "/Users/larkost/projects/rethinkdb-test/test/common/unit.py", line 51, in run
    subprocess.check_call([self.unit_executable, "--gtest_filter=" + filter])
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['/Users/larkost/projects/rethinkdb-test/build/debug_clang/rethinkdb-unittest', '--gtest_filter=ClusteringBranch.*']' returned non-zero exit status 1

Command '['/Users/larkost/projects/rethinkdb-test/build/debug_clang/rethinkdb-unittest', '--gtest_filter=ClusteringBranch.*']' returned non-zero exit status 1
```

**Comment by danielmewes** [comment]
There are some `nap(100)`s in this test, so there's a chance that this is just timing sensitive. I'll reduce those `nap`s and check if I can reproduce the same failure more reliably that way.

**Comment by danielmewes** [comment]
Never mind that. This appears to happen independently of how I set those nap times.

**Comment by danielmewes** [comment]
This is starting to look like a real race condition in the backfilling code (unless it's only broken because the mock store behaves incorrectly).

**Comment by danielmewes** [comment]
After further investigation, the problem appears to be a missing `on_empty_range` in the mock store's `send_backfill()` implementation.
Remaining steps:
* Document the issue properly, to make this easier to figure out in the future, and to stop us from making the same mistake again
* Double-check that the btree-backed store behaves correctly
* Check if we can add any verification steps to catch such problems in the future

**Comment by danielmewes** [comment]
I forgot some of the details since I've worked on this, but the bug fix (without additional verification code) is now up in review 3458.

**Comment by danielmewes** [comment]
Fixed in 8c9b554 on v2.2.x and b6e2a20 on next.
